File: backend/app/models/supplychain.py
"""Supply Chain models — suppliers, procurement requisitions, GRNs, supplier returns."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


# ── Supplier ─────────────────────────────────────────────────────────────────
class Supplier(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Registered supplier / vendor."""

    __tablename__ = "sc_suppliers"

    name: Mapped[str] = mapped_column(String(200), nullable=False)
    code: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    contact_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    email: Mapped[str | None] = mapped_column(String(200), nullable=True)
    phone: Mapped[str | None] = mapped_column(String(50), nullable=True)
    address: Mapped[str | None] = mapped_column(Text, nullable=True)
    payment_terms: Mapped[str | None] = mapped_column(
        String(20), nullable=True
    )  # net30, net60, cod
    payment_terms_days: Mapped[int] = mapped_column(Integer, default=30)
    rating: Mapped[int | None] = mapped_column(Integer, nullable=True)  # 1-5
    tags: Mapped[list[str] | None] = mapped_column(ARRAY(String), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    contact_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("crm_contacts.id"), nullable=True
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    owner_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )


# ── ProcurementRequisition ───────────────────────────────────────────────────
class ProcurementRequisition(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Internal purchase request before a PO is created."""

    __tablename__ = "sc_requisitions"

    requisition_number: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    requested_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    department_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("hr_departments.id"), nullable=True
    )
    status: Mapped[str] = mapped_column(
        String(30), default="draft"
    )  # draft, submitted, approved, rejected, converted_to_po
    approved_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    priority: Mapped[str] = mapped_column(String(20), default="medium")  # low, medium, high, urgent
    required_by_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    total_estimated: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    lines = relationship(
        "RequisitionLine", back_populates="requisition", cascade="all, delete-orphan"
    )


# ── RequisitionLine ──────────────────────────────────────────────────────────
class RequisitionLine(UUIDPrimaryKeyMixin, Base):
    """Single line on a procurement requisition."""

    __tablename__ = "sc_requisition_lines"

    requisition_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sc_requisitions.id"), nullable=False
    )
    item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_items.id"), nullable=False
    )
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    estimated_unit_price: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0)
    supplier_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sc_suppliers.id"), nullable=True
    )
    notes: Mapped[str | None] = mapped_column(String(500), nullable=True)

    requisition = relationship("ProcurementRequisition", back_populates="lines")
    item = relationship("InventoryItem")
    supplier = relationship("Supplier")


# ── GoodsReceivedNote ────────────────────────────────────────────────────────
class GoodsReceivedNote(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Goods received note — records delivery against a purchase order."""

    __tablename__ = "sc_goods_received_notes"

    grn_number: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    purchase_order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_purchase_orders.id"), nullable=False
    )
    supplier_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sc_suppliers.id"), nullable=False
    )
    warehouse_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_warehouses.id"), nullable=False
    )
    received_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    received_date: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[str] = mapped_column(
        String(20), default="draft"
    )  # draft, inspecting, accepted, partial, rejected
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    lines = relationship(
        "GRNLine", back_populates="grn", cascade="all, delete-orphan"
    )
    purchase_order = relationship("PurchaseOrder")
    supplier = relationship("Supplier")
    warehouse = relationship("Warehouse")


# ── GRNLine ──────────────────────────────────────────────────────────────────
class GRNLine(UUIDPrimaryKeyMixin, Base):
    """Single line on a GRN — tracks ordered vs received vs accepted."""

    __tablename__ = "sc_grn_lines"

    grn_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sc_goods_received_notes.id"), nullable=False
    )
    po_line_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_purchase_order_lines.id"), nullable=False
    )
    item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_items.id"), nullable=False
    )
    ordered_quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    received_quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    accepted_quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    rejected_quantity: Mapped[int] = mapped_column(Integer, default=0)
    rejection_reason: Mapped[str | None] = mapped_column(String(500), nullable=True)

    grn = relationship("GoodsReceivedNote", back_populates="lines")
    item = relationship("InventoryItem")


# ── SupplierReturn ───────────────────────────────────────────────────────────
class SupplierReturn(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Return of goods back to a supplier."""

    __tablename__ = "sc_supplier_returns"

    return_number: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    supplier_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sc_suppliers.id"), nullable=False
    )
    grn_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sc_goods_received_notes.id"), nullable=True
    )
    warehouse_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_warehouses.id"), nullable=False
    )
    status: Mapped[str] = mapped_column(
        String(30), default="draft"
    )  # draft, pending_approval, approved, shipped, completed, cancelled
    reason: Mapped[str] = mapped_column(Text, nullable=False)
    total_value: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    created_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )

    lines = relationship(
        "SupplierReturnLine", back_populates="supplier_return", cascade="all, delete-orphan"
    )
    supplier = relationship("Supplier")
    warehouse = relationship("Warehouse")


# ── SupplierReturnLine ───────────────────────────────────────────────────────
class SupplierReturnLine(UUIDPrimaryKeyMixin, Base):
    """Single line on a supplier return."""

    __tablename__ = "sc_supplier_return_lines"

    return_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sc_supplier_returns.id"), nullable=False
    )
    item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_items.id"), nullable=False
    )
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    unit_cost: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    reason: Mapped[str | None] = mapped_column(String(500), nullable=True)

    supplier_return = relationship("SupplierReturn", back_populates="lines")
    item = relationship("InventoryItem")

File: frontend/src/api/supplychain.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import apiClient from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface Supplier {
  id: string
  name: string
  code: string
  contact_name: string | null
  email: string | null
  phone: string | null
  address: string | null
  payment_terms: string | null
  payment_terms_days: number
  rating: number | null
  tags: string[] | null
  is_active: boolean
  contact_id: string | null
  notes: string | null
  owner_id: string
  created_at: string
  updated_at: string
}

export interface RequisitionLine {
  id: string
  requisition_id: string
  item_id: string
  quantity: number
  estimated_unit_price: number
  supplier_id: string | null
  notes: string | null
}

export interface ProcurementRequisition {
  id: string
  requisition_number: string
  title: string
  description: string | null
  requested_by: string
  department_id: string | null
  status: string
  approved_by: string | null
  approved_at: string | null
  priority: string
  required_by_date: string | null
  total_estimated: number
  notes: string | null
  created_at: string
  updated_at: string
}

export interface RequisitionDetail extends ProcurementRequisition {
  lines: RequisitionLine[]
}

export interface GRNLine {
  id: string
  grn_id: string
  po_line_id: string
  item_id: string
  ordered_quantity: number
  received_quantity: number
  accepted_quantity: number
  rejected_quantity: number
  rejection_reason: string | null
}

export interface GoodsReceivedNote {
  id: string
  grn_number: string
  purchase_order_id: string
  supplier_id: string
  warehouse_id: string
  received_by: string
  received_date: string
  status: string
  notes: string | null
  created_at: string
  updated_at: string
}

export interface GRNDetail extends GoodsReceivedNote {
  lines: GRNLine[]
}

export interface SupplierReturnLine {
  id: string
  return_id: string
  item_id: string
  quantity: number
  unit_cost: number
  reason: string | null
}

export interface SupplierReturn {
  id: string
  return_number: string
  supplier_id: string
  grn_id: string | null
  warehouse_id: string
  status: string
  reason: string
  total_value: number
  created_by: string
  created_at: string
  updated_at: string
}

export interface SupplierReturnDetail extends SupplierReturn {
  lines: SupplierReturnLine[]
}

export interface SupplyChainStats {
  total_suppliers: number
  pending_requisitions: number
  open_grns: number
  pending_returns: number
  pending_requisition_value: string
  pending_return_value: string
}

// ─── Paginated response types ─────────────────────────────────────────────────

export interface PaginatedSuppliers {
  total: number
  suppliers: Supplier[]
}

export interface PaginatedRequisitions {
  total: number
  requisitions: ProcurementRequisition[]
}

export interface PaginatedGRNs {
  total: number
  grns: GoodsReceivedNote[]
}

export interface PaginatedReturns {
  total: number
  returns: SupplierReturn[]
}

// ─── Payload types ────────────────────────────────────────────────────────────

export interface CreateSupplierPayload {
  name: string
  contact_name?: string
  email?: string
  phone?: string
  address?: string
  payment_terms?: string
  payment_terms_days?: number
  rating?: number
  tags?: string[]
  contact_id?: string
  notes?: string
}

export interface UpdateSupplierPayload extends Partial<CreateSupplierPayload> {
  id: string
  is_active?: boolean
}

export interface RequisitionLineIn {
  item_id: string
  quantity: number
  estimated_unit_price: number
  supplier_id?: string
  notes?: string
}

export interface CreateRequisitionPayload {
  title: string
  description?: string
  department_id?: string
  priority?: string
  required_by_date?: string
  notes?: string
  lines: RequisitionLineIn[]
}

export interface UpdateRequisitionPayload extends Partial<Omit<CreateRequisitionPayload, 'lines'>> {
  id: string
  lines?: RequisitionLineIn[]
}

export interface GRNLineIn {
  po_line_id: string
  item_id: string
  ordered_quantity: number
  received_quantity: number
  accepted_quantity: number
  rejected_quantity?: number
  rejection_reason?: string
}

export interface CreateGRNPayload {
  purchase_order_id: string
  supplier_id: string
  warehouse_id: string
  received_date: string
  notes?: string
  lines: GRNLineIn[]
}

export interface ReturnLineIn {
  item_id: string
  quantity: number
  unit_cost: number
  reason?: string
}

export interface CreateReturnPayload {
  supplier_id: string
  grn_id?: string
  warehouse_id: string
  reason: string
  lines: ReturnLineIn[]
}

// ─── Suppliers ────────────────────────────────────────────────────────────────

export function useSuppliers(params: {
  search?: string
  is_active?: boolean
  skip?: number
  limit?: number
} = {}) {
  return useQuery({
    queryKey: ['supplychain', 'suppliers', params],
    queryFn: async () => {
      const { data } = await apiClient.get<PaginatedSuppliers>('/supply-chain/suppliers', { params })
      return data
    },
  })
}

export function useSupplier(id: string) {
  return useQuery({
    queryKey: ['supplychain', 'suppliers', id],
    queryFn: async () => {
      const { data } = await apiClient.get<Supplier>(`/supply-chain/suppliers/${id}`)
      return data
    },
    enabled: !!id,
  })
}

export function useCreateSupplier() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CreateSupplierPayload) => {
      const { data } = await apiClient.post<Supplier>('/supply-chain/suppliers', payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'suppliers'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useUpdateSupplier() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, ...payload }: UpdateSupplierPayload) => {
      const { data } = await apiClient.put<Supplier>(`/supply-chain/suppliers/${id}`, payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'suppliers'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useDeleteSupplier() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      await apiClient.delete(`/supply-chain/suppliers/${id}`)
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'suppliers'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

// ─── Requisitions ─────────────────────────────────────────────────────────────

export function useRequisitions(params: {
  status?: string
  priority?: string
  skip?: number
  limit?: number
} = {}) {
  return useQuery({
    queryKey: ['supplychain', 'requisitions', params],
    queryFn: async () => {
      const { data } = await apiClient.get<PaginatedRequisitions>('/supply-chain/requisitions', { params })
      return data
    },
  })
}

export function useRequisition(id: string) {
  return useQuery({
    queryKey: ['supplychain', 'requisitions', id],
    queryFn: async () => {
      const { data } = await apiClient.get<RequisitionDetail>(`/supply-chain/requisitions/${id}`)
      return data
    },
    enabled: !!id,
  })
}

export function useCreateRequisition() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CreateRequisitionPayload) => {
      const { data } = await apiClient.post<RequisitionDetail>('/supply-chain/requisitions', payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'requisitions'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useUpdateRequisition() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, ...payload }: UpdateRequisitionPayload) => {
      const { data } = await apiClient.put<RequisitionDetail>(`/supply-chain/requisitions/${id}`, payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'requisitions'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useSubmitRequisition() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await apiClient.post<ProcurementRequisition>(`/supply-chain/requisitions/${id}/submit`)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'requisitions'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useApproveRequisition() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, action }: { id: string; action: 'approve' | 'reject' }) => {
      const { data } = await apiClient.post<ProcurementRequisition>(
        `/supply-chain/requisitions/${id}/approve`,
        null,
        { params: { action } }
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'requisitions'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useConvertRequisitionToPO() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await apiClient.post<{ message: string; purchase_order_id: string; po_number: string }>(
        `/supply-chain/requisitions/${id}/convert-to-po`
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'requisitions'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
      qc.invalidateQueries({ queryKey: ['inventory'] })
    },
  })
}

// ─── GRNs ─────────────────────────────────────────────────────────────────────

export function useGRNs(params: {
  status?: string
  skip?: number
  limit?: number
} = {}) {
  return useQuery({
    queryKey: ['supplychain', 'grns', params],
    queryFn: async () => {
      const { data } = await apiClient.get<PaginatedGRNs>('/supply-chain/grn', { params })
      return data
    },
  })
}

export function useGRN(id: string) {
  return useQuery({
    queryKey: ['supplychain', 'grns', id],
    queryFn: async () => {
      const { data } = await apiClient.get<GRNDetail>(`/supply-chain/grn/${id}`)
      return data
    },
    enabled: !!id,
  })
}

export function useCreateGRN() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CreateGRNPayload) => {
      const { data } = await apiClient.post<GRNDetail>('/supply-chain/grn', payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'grns'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useAcceptGRN() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await apiClient.post<GoodsReceivedNote>(`/supply-chain/grn/${id}/accept`)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'grns'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
      qc.invalidateQueries({ queryKey: ['inventory'] })
    },
  })
}

export function useRejectGRN() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await apiClient.post<GoodsReceivedNote>(`/supply-chain/grn/${id}/reject`)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'grns'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

// ─── Returns ──────────────────────────────────────────────────────────────────

export function useReturns(params: {
  status?: string
  skip?: number
  limit?: number
} = {}) {
  return useQuery({
    queryKey: ['supplychain', 'returns', params],
    queryFn: async () => {
      const { data } = await apiClient.get<PaginatedReturns>('/supply-chain/returns', { params })
      return data
    },
  })
}

export function useReturn(id: string) {
  return useQuery({
    queryKey: ['supplychain', 'returns', id],
    queryFn: async () => {
      const { data } = await apiClient.get<SupplierReturnDetail>(`/supply-chain/returns/${id}`)
      return data
    },
    enabled: !!id,
  })
}

export function useCreateReturn() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CreateReturnPayload) => {
      const { data } = await apiClient.post<SupplierReturnDetail>('/supply-chain/returns', payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'returns'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useApproveReturn() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await apiClient.post<SupplierReturn>(`/supply-chain/returns/${id}/approve`)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'returns'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
    },
  })
}

export function useCompleteReturn() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await apiClient.post<SupplierReturn>(`/supply-chain/returns/${id}/complete`)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['supplychain', 'returns'] })
      qc.invalidateQueries({ queryKey: ['supplychain', 'dashboard'] })
      qc.invalidateQueries({ queryKey: ['inventory'] })
    },
  })
}

// ─── Dashboard ────────────────────────────────────────────────────────────────

export function useSupplyChainDashboard() {
  return useQuery({
    queryKey: ['supplychain', 'dashboard', 'stats'],
    queryFn: async () => {
      const { data } = await apiClient.get<SupplyChainStats>('/supply-chain/dashboard/stats')
      return data
    },
  })
}
